[#33906] New Media Manager for Joomla! CMS

Buddhima opened the pull request:
Hi All,
Let me introduce the new media manager for Joomla! along with this PR. With this project I have introduced 12 new things to Media Manager. I have constructed a document describing this project in detail.
The link for the document is : https://docs.google.com/document/d/1mZMwBU_hW7AshL5rkAdIyswQmzBmP1L8L1PdM093rB0/edit?usp=sharing
Feature Tracker Item: http://joomlacode.org/gf/project/joomla/tracker/?action=TrackerItemEdit&tracker_item_id=33906&start=0
This new Media Manager add entries to your database. Therefore I kindly request you to go through "Testing Instructions" at document before testing this.

Thank You!

![image](https://cloud.githubusercontent.com/assets/1543384/7898842/d0e5d650-072a-11e5-9be6-b3e027606709.png)

Buddhima committed:
Creating folder structure

Buddhima committed:
Setting up models and views of new MVC

Buddhima committed:
setup media entry point to direct to new MVC

Buddhima committed:
small changes in mvc

Buddhima committed:
Make media and media list to display

Buddhima committed:
Folder navigation checking issue fixed

Buddhima committed:
Add new folder functionality working

Buddhima committed:
Implemented Delete for folders and files in browser component

Buddhima committed:
File uploading functionality working

Buddhima committed:
Initialize Media Manager Editor

Buddhima committed:
Cropping image working

Buddhima committed:
Renaming image works

Buddhima committed:
Resizing image works

Buddhima committed:
Rotating image works

Buddhima committed:
Filterig images work

Buddhima committed:
Editor cancel operation implemented

Buddhima committed:
View ACL setted

Buddhima committed:
Turn Media component buttons to show modals

Buddhima committed:
Cancel to media/cancel.php

Buddhima committed:
Bring back layouts for buttons, change JLayoutFile to use /layout folder

Buddhima committed:
JQuery for image preview and lightbox back to work

Buddhima committed:
Thumbs creation added

Buddhima committed:
Modify ftp access form, ftpValidate controller and upload controller

Buddhima committed:
Drag&Drop file-uploader added

Buddhima committed:
Drag&Drop modifications-1

Buddhima committed:
Resizing hight-width auto scaling added

Buddhima committed:
Thumbs in medialist resized

Buddhima committed:
Modifications to thumbs-2

Buddhima committed:
Fixed Details view to direct Editor

Buddhima committed:
Fixed issue with thumbs_folder checkbox

Buddhima committed:
MediaTableMedia initiated

Buddhima committed:
Commiting cms/model from Elin's repo

Buddhima committed:
Show basic information on Editor view

Buddhima committed:
Table updating added to models

Buddhima committed:
Hashed filename when uploading

Buddhima committed:
Initial work related to editor's image form

Buddhima committed:
Work related to editor's image form - 2

Buddhima committed:
Work related to editor's image form - 3

Buddhima committed:
Create controller added

Buddhima committed:
Remove #__media table

Buddhima committed:
Removed 'Rename' action from editor

Buddhima committed:
Media manager adds which are not in table & core_content_id appended to
image link

Buddhima committed:
Make Editor to work with ID and modify models

Buddhima committed:
Add 'id' in detail mode and change to 'id' in thumbnail mode

Buddhima committed:
Removing JObject from models and working on Tags - 1

Buddhima committed:
working on Tags - 2

Buddhima committed:
Fixing issue with model of Medialist, modify treeview and remove file
name on top in Editor

Buddhima committed:
working on Tags - 3

Buddhima committed:
working on Tags - 4

Buddhima committed:
Added 'path' filter where folder location is taken from input

Buddhima committed:
Fixed batch delete issue

Buddhima committed:
Checkin, Checkout functioning

Buddhima committed:
Get type_id from content_type table, modify image.xml form

Buddhima committed:
Work on Media Manager-Popup. Create classes, make it to do basics
(display etc.)

Buddhima committed:
Working on Media Manager Pop-up 2

Buddhima committed:
Media Manager - popup directory list is functioning as expected

Buddhima committed:
Fixed issue with displaying image information

Buddhima committed:
File Browser of popup modified

Buddhima committed:
Fixed popup image manager ui issues, fixes in file uploading

Buddhima committed:
Sidebar added and moved Tree View to sidebar

Buddhima committed:
category section added to access xml of media

Buddhima committed:
Adding actions to enable operations in category

Buddhima committed:
Conflict Solve in MediaManager.js

Buddhima committed:
Fixed issue with making jQuery work - (properly substitute
getElementById by jQuery)

Buddhima committed:
Modify D&D js to suite jQuery (use 'jQuery' instead of $)

Buddhima committed:
small fix: image-popup , up button resized

Buddhima committed:
Urls save as relative paths from JPATH_ROOT

Buddhima committed:
Incorporate latest changes to com_media

Buddhima committed:
Move actions to toolbar (except crop)

Buddhima committed:
Added 'Crop' toolbar-button & completely eleminate Actions tmpl
(tmpl/default_actions.php)

Buddhima committed:
Fixed the bug - not showing MediaManager popup in the frontend

Buddhima committed:
Code cleanup - 1

Buddhima committed:
Code cleanup - 2

Buddhima committed:
Code cleanup - 3 (Established backward compatability)

Buddhima committed:
Code cleanup - 4 (Specially language strings updated)

Buddhima committed:
Solve issue with path by using DIRECTORY_SEPARATOR

Buddhima committed:
Changes in language files and DB files added, Thanks JM

Buddhima committed:
Changes in language files and 'Uncategorised' category added

Buddhima committed:
Fixing folder icon size in media manager

Buddhima committed:
Hashing Cancelled, Drag&Drop Uploding error handling fixed, Image
Selector fixed, coding style issues fixed

Buddhima committed:
Bring back legacy models in com_media with deprecated warning, File
override option, Delete confermation to toolbar button, Language
translations, Jcrop gif, popups resized, and new toolbar icons in editor

Buddhima committed:
Solved issue with saving and removing b/e tags

Buddhima committed:
Solved issue with saving and removing b/e tags - 2 & putting
core_content_item_id in #_ucm_content

Buddhima committed:
Adding frontend display images when linking from tags

Buddhima committed:
Updated sql files with #__content_types -> table, field mappings

Buddhima committed:
Make FTP back to work in Media Manager

Buddhima committed:
upload prepending, slash after foldername, and make session expiration
an error

Buddhima committed:
Code cleanup - 5 (Using PHPCS on most MVC classes)

Buddhima committed:
gracefully handled post size exceeding and slice file name to fit in
uploader

Buddhima committed:
Merge branch 'master' into new_com_media

Conflicts:
	administrator/components/com_media/controllers/file.php
	administrator/language/en-GB/en-GB.com_media.ini

Conflicts solved

Buddhima committed:
Fixing issue with popup uploader (showing space charanters)

Buddhima committed:
Fixing issue with sql files

Buddhima committed:
Merge branch 'master' into new_com_media

Conflicts:
	administrator/components/com_media/controller.php
	administrator/components/com_media/controllers/file.php
	administrator/components/com_media/controllers/folder.php
	administrator/components/com_media/helpers/media.php
	administrator/components/com_media/media.php
	administrator/components/com_media/models/list.php
	administrator/components/com_media/models/manager.php
	administrator/components/com_media/views/images/tmpl/default.php
	administrator/components/com_media/views/images/view.html.php
	administrator/components/com_media/views/imageslist/tmpl/default.php
	administrator/components/com_media/views/imageslist/tmpl/default_image.php
	administrator/components/com_media/views/imageslist/view.html.php
	administrator/components/com_media/views/media/view.html.php
	administrator/components/com_media/views/medialist/tmpl/details_doc.php
	administrator/components/com_media/views/medialist/tmpl/details_folder.php
	administrator/components/com_media/views/medialist/tmpl/details_img.php
	administrator/components/com_media/views/medialist/tmpl/details_up.php
	administrator/components/com_media/views/medialist/tmpl/thumbs_doc.php
	administrator/components/com_media/views/medialist/tmpl/thumbs_folder.php
	administrator/components/com_media/views/medialist/tmpl/thumbs_img.php
	administrator/components/com_media/views/medialist/tmpl/thumbs_up.php
	administrator/components/com_media/views/medialist/view.html.php
	libraries/cms/form/field/media.php

Buddhima committed:
Styles issue fixed in components/com_media/media.php

Buddhima committed:
Fixing Drag & Drop uploader style issue

Buddhima committed:
Changing the thumbnail design

Buddhima committed:
Modify thumbnails along with JS and CSS for other types

Buddhima committed:
Solving style issues with editor-xtd/image/image.php

Buddhima committed:
Enable uploading files with non-alphanumeric and space characters

Buddhima committed:
Use punycode and modify filename transform

Buddhima committed:
Fixing issue occoured because of new mutable sidebar

Buddhima committed:
A change in filename field of uploader

Buddhima committed:
Adding media copy and move functionality

Buddhima committed:
Enabling folders to copy and move

Buddhima committed:
Merge branch 'master' into new_com_media

Conflicts:
	administrator/components/com_media/views/media/tmpl/default.php

Buddhima committed:
Code style fix in copy.php file

bembelimen commented on administrator/components/com_media/controller/editor/crop.php:
```diff
@@ -0,0 +1,73 @@
+
+		$model = new $modelClass;
+
+		if (empty($w) && empty($h) && empty($x) && empty($y))
```
Hello,
first of all, I really like your commit and this project, good work!
But when I looked into the code, I wondered, about the request parameter. In fact I miss a good validation for this stuff.
For example: The ID should be an INT (getInt('id')) and the $x, $y, $width, $height should have some valid value (> 0 I guess for the width/height) etc.
Also I wondered, if line 52 here make sense? $width/$height are required, but $x, $y not (I think) so wouldn't make something like:
```php
if (empty($w) || empty($h))
```
more sense?

I saw a lot of places in your commit, where you had such small "issues". Nothing serious, but I think something you should take care off...

Buddhima replied:
Hi @bembelimen ,
Thank you for looking into this project. I'll check on the issues, and let me know if there are any others too.
Thank You!

bembelimen replied:
Hi,
is there a way to point at several lines with one comment? Otherwise it could become a little bit "spammy"

Buddhima replied:
Hi @bembelimen ,
If those lines are next to each other, showing them as in previous comment is enough. Else need to specify separately. It won't be a spammy because, once those are fixed note get collapsed by-default. 
Thanks!

zero-24 replied:
@bembelimen 

>is there a way to point at several lines with one comment? Otherwise it could become a little bit "spammy"

Yes. do one comment here and link to all places that you would comment e.g.:
https://github.com/joomla/joomla-cms/pull/3839/files#diff-fb4315da042aefbbc6dc15cc0108e5f6R52

You can gat the links by clicking the line number on the files (https://github.com/joomla/joomla-cms/pull/3839/files) part of this PR :smile:

zero-24 commented:
@Buddhima this looks out of sync. Can you resync?

maybe we need also to apply `Media manager: hide the Up button when in base directory` https://github.com/joomla/joomla-cms/pull/5704

Buddhima committed:
Merge branch 'master' into new_com_media

Conflicts:
	administrator/components/com_media/layouts/toolbar/deletemedia.php
	administrator/components/com_media/layouts/toolbar/newfolder.php
	administrator/components/com_media/layouts/toolbar/uploadmedia.php
	administrator/components/com_media/views/imageslist/tmpl/default_folder.php
	administrator/components/com_media/views/media/tmpl/default_folders.php
	administrator/components/com_media/views/media/tmpl/default_navigation.php
	administrator/components/com_media/views/medialist/tmpl/default.php
	administrator/components/com_media/views/medialist/tmpl/details.php
	administrator/components/com_media/views/medialist/tmpl/thumbs.php
	tests/unit/suites/libraries/joomla/document/raw/JDocumentRawTest.php
	tests/unit/suites/libraries/joomla/document/xml/JDocumentXmlTest.php

Buddhima committed:
Hide up button in base directory & fixed side bar

Buddhima committed:
Update copyrights and remove index.html files in com_media

Buddhima commented:
Hi @zero-24 , 
Now this is synced back and included the mentioned change.
Thanks!

zero-24 commented:
sorry @Buddhima just one hour ago @infograf768 merge another fix for com_media https://github.com/joomla/joomla-cms/pull/5722 can you have a look into it?

If i get some free time i will test it. Thank you so much for your work here!!

Buddhima committed:
Displaying the path to oped folder - PR #5722

Buddhima committed:
Merge branch 'master' into new_com_media

Buddhima commented:
Hi @zero-24 ,
I added changes in given PR .Thank you for pointing it and appreciate for paying your attention on this.

zero-24 commented on administrator/components/com_media/controller/editor/save.php:
```diff
@@ -0,0 +1,102 @@
+		$item = $model->getItem($id);
+
+		// Replace array with non-existing properties
+// 		$data = array_replace_recursive((array) $item, $data);
```
can this be removed or it this temp?

zero-24 commented:
Hi,

here are some results / issues 

![media_create_folder_message_missing_folder](https://cloud.githubusercontent.com/assets/2596554/5788737/9fccc286-9e48-11e4-86a8-09c6c3aaf32a.PNG)
![media_delete_folder_message_missing_folder](https://cloud.githubusercontent.com/assets/2596554/5788740/9fdb40f4-9e48-11e4-974a-35d3861b89fe.PNG)

The create and delete message don't include that the folder was deleted or added.

![media_delet_without_selection](https://cloud.githubusercontent.com/assets/2596554/5788739/9fd6b4bc-9e48-11e4-818b-13cacbcda029.PNG)

if you press the delete button without make a selection you get this view.

![media_create_folder_windows](https://cloud.githubusercontent.com/assets/2596554/5788738/9fd12844-9e48-11e4-8896-5f1aa4f24cd3.PNG)

The create folder modal on windows. The folder looks a bit messed up. 

![media_images](https://cloud.githubusercontent.com/assets/2596554/5788741/9fdc1498-9e48-11e4-853f-e2dab0b12c99.PNG)

Can we make the images better ti overlab with the select and delet buttons.

![media_sidebar](https://cloud.githubusercontent.com/assets/2596554/5788742/9fe6a1e2-9e48-11e4-99db-c30a28f76027.PNG)

Can we have a look into the sidebar? It looks we have min a line to much here.

![media_thubs_folder_start_point_rel_or_absolute_examples_description](https://cloud.githubusercontent.com/assets/2596554/5788743/9fea30be-9e48-11e4-8142-6d76262e2cfc.PNG)

Can we add here a decription that explains what is the stating point?
`\www\...` or `\images\... or `\banners\` etc. 

![media_uploader](https://cloud.githubusercontent.com/assets/2596554/5788761/833050d8-9e49-11e4-936b-002311c4fd78.PNG)

can we make the uploader / the progressbars better? The successful is  very smal and the errord have not the same length.
maybe we can make the progressbars a bit bigger and with rounded.

I will do more testing if i get time again

zero-24 commented on administrator/components/com_media/controller.php:
```diff
@@ -12,7 +12,10 @@
  * Media Manager Component Controller
  *
- * @since  1.5
+ * @package     Joomla.Administrator
```
The tags @package and @subpackage can be removed form the class definition.
+ * @package     Joomla.Administrator
+ * @subpackage  com_media

Buddhima replied:
Removing ```@package``` gives error ``` Missing @package tag in class comment ```

mbabker replied:
You probably need to merge the current staging branch into your working branch then.  IIRC it was after 3.3.6 was released that we relaxed the requirement for this in the PHPCS ruleset.

Buddhima replied:
Thank you @mbabker . I removed those from the component.

zero-24 commented:
@Buddhima 

can you add the default category `uncategorised` category to the install sql files?

maybe we should use `Uncategorised` instead of `uncategorised` @brianteeman ?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/3839">issues.joomla.org/joomla-cms/3839</a>.</sub>

Buddhima commented:
Hi @zero-24 , 
I already have added that : https://github.com/Buddhima/joomla-cms/blob/new_com_media/administrator/components/com_admin/sql/updates/mysql/3.5.0-2014-06-28.sql#L4
Thank You!

zero-24 commented:
>I already have added that : https://github.com/Buddhima/joomla-cms/blob/new_com_media/administrator/components/com_admin/sql/updates/mysql/3.5.0-2014-06-28.sql#L4

Yes. But this is only for updates :smile: 

We need it also here: 
https://github.com/joomla/joomla-cms/tree/staging/installation/sql

IIRC at the joomla.sql something like:
https://github.com/joomla/joomla-cms/blob/staging/installation/sql/mysql/joomla.sql#L240-L246

Buddhima committed:
Fixing few misc. ui issues: thumbnail, messages, progressbar & delete
redirection

Buddhima committed:
Merge branch 'master' into new_com_media

Conflicts:
	administrator/components/com_media/controllers/folder.php
	administrator/language/en-GB/en-GB.com_media.ini
	language/en-GB/en-GB.com_media.ini

Add delete empty error message

Buddhima commented:
Hi @zero-24 ,

Thank you for the feedback, it took a bit long period to have a look at those. So following are the main changes done so far.
1. Create & Delete messages for folders added separately.
2. Redirection after delete (false situation) is working fine with an error message
3. Create folder path fixed. Have to use / since all operating systems supports that. I appreciate if you can have a check on that since I'm currently working on Linux.
4. Thumbnails refactored once again.
5. Starting point to Create Thumbnails added.
6. Uploader progress bar always caused problems :( . Since now modal supports responsiveness, moved back to %-based separation

Also, I think there's no need to change ```uncatagorised``` term. It has no difference from other cases.

I think moving these to installation file should be done at the end, else this PR will get conflicts more frequently.

Would like to know your thoughts and experience with current implementation.

Thank you!

Buddhima committed:
Removing @package and @subpackage from class comment

losedk commented:
How can we move this awesome contribution by @Buddhima ahead?

roland-d commented:
@losedk Unfortunately this PR can't move ahead as it stands now because it is depending on other code changes that have not yet been decided upon and in addition it is extending from com_config which is not preferable.

Personally I would love to see this go into Joomla, hence all the testing I have done, but code-wise it isn't there yet. If we are to move ahead, the code changes should be updated to fit into the current codebase of Joomla.

Mathewlenning commented:
I need to put my 2 cents in to this conversation. So at least later on I can say I tried to prevent this.

The single task controller design in com_config and this component is flawed. 21 controllers to push 2 models with 4 views in the admin area is just not sustainable. 

Please do not merge this feature without carefully considering the conversations 6 months from now when the joy of a shiny new feature turns into the grind of maintaining another overly complex system.

I know a lot of people want this new feature, but I ask them to consider how useful it will be if every few minor releases it breaks unintentionally. Simply because in the 8000+ new lines, someone didn't realize changing X here would break Z there. 

@Buddhima I appreciate the hard work and initiative that it took to bring this much desired upgrade to the Joomla community. 

So please don't be offend by my comments. The architectural design is the recipient of my criticism. It is not directed at you or your code.

If you haven't heard me ranting on twitter or the google group, I've come up with an alternative approach to single task controllers. I'm calling it MVSC.

If you are interested in looking at the design check out the com_config rebuild, It isn't complete but it should give you a good idea of how MVSC is different from your approach.

https://github.com/icarus/logical-j/tree/com_config

Buddhima commented:
Hi @losedk , @roland-d , @Mathewlenning et al.,

I thought of raising my voice at this point since discussions flow seems forgetting the initial point.
In my comment 1st July, 2014, I have clearly stated that "After completing new MVC project, this also will need to refactor. "

New MVC in Joomla is an effort to introduce single task controllers. I have done it for com_config. As you have done with sample, in com_config refactoring, I initially put base classes to lib. But unfortunately have to take them out at the end. I know @wilsonge continued it and I also know @Mathewlenning  helped @wilsonge a lot during his project. 

Also I have mentioned in the doc that, this is ti stimulate the necessity of single task controllers (I refer as new MVC and @Mathewlenning refers as MVSC).

Finally what I need to see is that, when single task controllers in the Joomla! core, this need to be modified accordingly and merged to Joomla! CMS.

Thank you!

Chalkin commented:
Greetings from JUC-München we tested your new Mediamanager with 6 people.
We are happy with the new media manage and happy that you extend it - the current one really needs some work.
Thanks for the good work we hope that this patch will be released as soon as possible.

**Bugs:**
- After scrolling down (on a big image) the editor buttons (Save, Crop, Resize….) are not clickable anymore.

![screen shot 2015-03-14 at 11 58 34](http://issues.joomla.org/uploads/1/bb3e7af7b2fb9ee609840b36249d3f48.png)

- After uploading files, the file extension should always be lower case - and if someone uploads the file the first time, the media manager should rename the file extension to lower case. Otherwise there will be issues on linux. (will not open the image on linux and Joomla will throw a JERROR)


**Suggestions:**

1) Rotate/Crop/Resize
All features should not overwrite the original image while live editing the file but store a copy of the file.

_**Example:**
I uploaded a picture, and rotate it now. While i click the rotate button it overwrites my original file. Which is not good._

It basicly should always keep the original file and should store the changed image as a new version (as a copy).
Also while editing the image chan on the fly should not change the file in the file system. It should save the (new copy) as soon as I apply/save the changes of my image.


2) Image Thumbnail Folder
The thumbnail folder should be a (global) option of com_media. Therefore we don’t have to type the folderpath every time.


3) Rotating
Add Buttons for: „90-Degree left“, „90-Degree right“, „Flip horizontaly“, „Flip verticaly"

**Usability enhancements:**
For the usebility approvments it would be nice to add folowing features:
The options for the Rotate/Crop/Resize tools are currently opend in a modal box. It would be better to implement them differently for better usability:

_**Option 1:**_
Options for the Buttons will be show as slideout/dropdown after clicking the button like it is the case in com_articles list when you click on „Search Tools“

![screen shot 2015-03-14 at 11 58 32](http://issues.joomla.org/uploads/1/360b283ef15159a844f9d3874ad65990.jpg)

_**Option 2:**_
Use „Tabs“ for the Buttons and place the Image on the left side and the options for the tools on the right sidebar 

![screen shot 2015-03-14 at 11 58 31](http://issues.joomla.org/uploads/1/ed20f101b1eee763ea2afb7d3bfcebee.jpg)


**Optional for better usability you can take a look at this script [See here](https://github.com/imazen/studiojs "see here")**<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/3839">issues.joomla.org/joomla-cms/3839</a>.</sub>

N6REJ commented:
NO NO NO.. do NOT make a copy with every edit.. good grief man!!! you 
could easily wind up with 100 copies of the same damn file!!
when you hit SAVE you save OVER the original file.  and you don't touch 
that file UNTIL the user has hit save.
if a temp file is needed for edits, thats fine, but don't overwrite the 
original until "save"
Bear
On 3/14/2015 12:03, Patrick wrote:
>
> Greetings from JUC-München we tested your new Mediamanager with 6 people.
> We are happy with the new media manage and happy that you extend it - 
> the current one really needs some work.
> Thanks for the good work we hope that this patch will be released as 
> soon as possible.
>
> *Suggestions:*
>
> 1) Rotate/Crop/Resize
> All features should not overwrite the original image while live 
> editing the file but store a copy of the file.
>
> /*Example:*
> I uploaded a picture, and rotate it now. While i click the rotate 
> button it overwrites my original file. Which is not good./
>
> It basicly should always keep the original file and should store the 
> changed image as a new version (as a copy).
> Also while editing the image chan on the fly should not change the 
> file in the file system. It should save the (new copy) as soon as I 
> apply/save the changes of my image.
>
>

Chalkin commented:
@N6REJ  Maybe you missunderstood us, or i missunderstand you.

**But that's how it should work in our opinion:**
1. You upload an image (original file)
    It saves this one. (just like it always does)
2. You edit the file (rotate it) and ONCE you click save it stores a copy and NOT right after you edit the file.
3. You edit it again, you overwrite the first/manipulated copy. So the copy is always the current, but changed version. And this way you always have the original file backed up. And you don't get 100s of copies!

_Why should it be like this?_
Well, you ALWAYS need a copy of the original file for being able to restore the changes you did. Let's say the user edits his uploaded image - for example crops it - and then the media manager overwrites the changed file right away and replaces the original file (like it is currently the case) - then you can NEVER go back to the original file. That is just bad in terms of usability.

You NEVER NEVER NEVER do that. Not in Photoshop (keyword: nondestructive/descructive workflow) nor in any other (web) based image editor. Every serios image editor keeps the original file so the user can always restore to it's original state. (in case of photoshop the user should work nondestructive)

**Current state (Bad):**
The user changes his uploaded file, for example crops it and the original file is overwritten. He can NEVER go back to the original one.
_**Worse:**_
Currently the media manager saves every single change right AFTER you manipulated the file. You don't even have to click "Apply" or "Save". So once you croped your file, there is no chance in going back. You have to reupload the original file.

wilsonge commented:
+1 for this change proposed. It doesn't really make sense to overwrite the old file until the user clicks save

Buddhima committed:
Editor can undo its actions

Buddhima committed:
Adding rotate direction selection

Buddhima committed:
Merge branch 'master' into new_com_media

Buddhima committed:
Adding flip method to JImage

Buddhima committed:
Adding flip image functionality to Editor

micker commented:
hello some return about end user mode
1 i think we have too much view => it will be better to drag and drop for upload on main view without open popup
2 rezize option isn't realy good end user don't whant too typ zsize => it will be better to have a dropdown list with more commun size and custom value to display field for custom size (with a default value checked in params)
it realy a wonderfull work !!! very important adding for j3.5 i hope
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/3839">issues.joomla.org/joomla-cms/3839</a>.</sub>

Buddhima commented:
Hi All,

With the recent commits to this, I would like to introduce you 3 new things to this project.

1. Undo changes 
From here onwards, when you do change image (eg: crop, resize) through Editor, it won't affect the original image. Once you click Save, modified image will replace the original image. If you click Cancel, modified one will be discarded.
Talking in technical terms, the modified images will get saved under '/tmp/com_media' folder. So that will be replaced original or discarded on the interest.

2. Image flip functionality
This was a missing function in JImage. So I included it in a separate PR for JImage, and use that in here as well. Please refer PR #6571 for further clarifications.

3. Add image rotate direction selection & convert image extensions to lower-case on upload.

Thank you all of you @N6REJ , @micker & @Chalkin (for JUC-Munich) for testing this. I will try to fulfill requested features as much as possible.

I also like to see this in Joomla! 3.5 :innocent:

Thank you!

Buddhima committed:
Fixing issue with image modal with image field

Buddhima committed:
Sidebar heading style change

dgt41 commented on administrator/components/com_media/controller.php:
```diff
@@ -25,8 +26,9 @@ class MediaController extends JControllerLegacy
+	 * @deprecated  4.0
 	 */
-	public function display($cachable = false, $urlparams = false)
+	public function display($cachable = false, $urlparams = array())
```
@Buddhima Can you check the changes I made in PR #5871 ? 
The changes allow for overriding the views in frowned/backend, which is something crucial in order to have layout override for the media form field

zorroh commented:
i like new features - nice job! :)

one big issue right now is that there isn't a good solution where users can access only specified folder... generated user folder during registration would be also good option. this helps to prevent case where there is one monster heavy picture folder and users can find their files/photos easy.

or nice option would be also if user wants upload photo to article then in upload window user sees only upload tab, in second tab is my uploaded files and third tab is all folders (if admin allows it) ---same solution is in google docs :)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/3839">issues.joomla.org/joomla-cms/3839</a>.</sub>

Buddhima committed:
Merge branch 'master' into new_com_media

Conflicts:
	media/media/js/mediamanager.js

Buddhima committed:
Adding trailing slash for clarity in new-folder modal

Buddhima committed:
Adding support for video and audio, fixing a bug in mediamanager.js

Buddhima committed:
Fixing codestyle issues

Buddhima commented:
Hi All,

With recent commits I added video preview, and audio preview support for the new media manager. I believe it would help to provide better user experience.

**Video Preview :**
![video_preview](https://cloud.githubusercontent.com/assets/1543384/8259502/96f3873e-16da-11e5-9c8d-c35d619e45dd.png)

**Audio Preview :**
![audio_preview](https://cloud.githubusercontent.com/assets/1543384/8259534/c6670414-16da-11e5-9481-b322d899893a.png)

Thank you!

dgt41 commented:
@Buddhima 
Looking good! 
Can you take a look at #5871 and also incorporate the changes I made there in the file controller.php here?

Also you are using mootools modal here, we are trying to render everything with bootstrap. I can make a PR changing that, if you want...

Buddhima commented:
@dgt41 
I'll have a look at your implementation as well. I would appreciate if you can send a PR.
Thank you!

dgt41 commented:
@Buddhima will do that this weekend. Are there any other places where you use mootools modal?

Buddhima commented:
@dgt41 I don't think so, because I used bootstrap modals for Media Editor popups. For video & audio popups, they need to come-out-of iframe as well. At the moment I wasn't able to do with mootools.

danielsmink commented:
@Buddhima any idea when we can expect this to be merged? 3.5 or later? We need a more advanced Media manager for a project and this sounds exactly like what we need. But considering deadlines it might come to late.

Buddhima commented:
Hi @danielsmink , According to the discussions above, it's sad to say that still this is not planned for Joomla! 3.5. For your case, you can take this and modify accordingly.

N6REJ commented:
if its ready, why is it being stopped?
Bear
On 9/8/2015 02:58, Buddhima Wijeweera wrote:
>
> Hi @danielsmink <https://github.com/danielsmink> , According to the 
> discussions above, it's sad to say that still this is not planned for 
> Joomla! 3.5. For your case, you can take this and modify accordingly.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/joomla/joomla-cms/pull/3839#issuecomment-138467499>.
>

Bakual commented:
> if its ready, why is it being stopped?

It's not ready due to two main points. Those are code/architecture issues and not user interface issues.

It depends a lot on classes from com_config, which is a no-go.
It either has to

* wait for the "new MVC" classes @wilsonge wanted to write, and then be rewritten to extend those,
* rewite the whole thing to directly use the base classes for the "new MVC" in the library folders
* or rewrite the whole thing to use the legacy MVC

It also still uses the UCM tables to store the data, which I would prefer when it would not do that. UCM as it is of today will likely be changed or even removed with J4. So we would face a rewrite then again.

N6REJ commented:
damn, ok.

Bear
On 9/8/2015 09:32, Thomas Hunziker wrote:
>
>     if its ready, why is it being stopped?
>
> It's not ready due to two main points. Those are code/architecture 
> issues and not user interface issues.
>
> It depends a lot on classes from com_config, which is a no-go.
> It either has to
>
>   * wait for the "new MVC" classes @wilsonge
>     <https://github.com/wilsonge> wanted to write, and then be
>     rewritten to extend those,
>   * rewite the whole thing to directly use the base classes for the
>     "new MVC" in the library folders
>   * or rewrite the whole thing to use the legacy MVC
>
> It also still uses the UCM tables to store the data, which I would 
> prefer when it would not do that. UCM as it is of today will likely be 
> changed or even removed with J4. So we would face a rewrite then again.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/joomla/joomla-cms/pull/3839#issuecomment-138583320>.
>

roland-d commented:
Since this PR is never going to be merged as-is I am going to close it. @Buddhima thank you for your work and I will continue to work on trying to get features moved over into the Joomla core. 

Those of you who want to work further with this code can go to Buddhimas' repo at: https://github.com/Buddhima/joomla-cms/tree/new_com_media

Thanks everybody for the feedback.

holodyn commented:
Come on - obviously there are several considerations to this PR, but there is a tremendous amount of work put toward this product that would transform Joomla! usability.  Merge this thing and let the community contribute to the updates afterwards rather than killing the momentum.  Good job  @Buddhima.

micker commented:
yes  dont kill this important feature all users need it

javigomez commented:
Just to clarify. Is not a close forever. Is just a delay. We really appreciate the work done here and we all what to see it happening soon.

JoshuaLewis commented:
Any news on this?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/3839">issues.joomla.org/joomla-cms/3839</a>.</sub>

wilsonge commented:
We're working on porting the features back into the existing MVC system here https://github.com/joomla-projects/media-manager-improvement but progress at the moment is reasonably slow. We are aiming to have most of the things from this PR ported into that for 3.6 in the next month or so however

chrisdavenport commented:
The latest code for this is here: https://github.com/jissereitsma/media-manager-improvement

If you have a little time to spare, please test and provide feedback.  Thanks.

N6REJ commented:
why two separate repos?

brianteeman commented:
Why just have one when you can have two or more ;)

piotr-cz commented:
It may be late suggestion, but are there any reasons for not creating just an extension repo, similar to [joomla-extensions/weblinks](https://github.com/joomla-extensions/weblinks)?

chrisdavenport commented:
Because it's not a properly decoupled extension like weblinks.

piotr-cz commented:
From brief look at the diff, it looks like two changes preventing decoupling are:

- [libraries/cms/form/field/media.php](https://github.com/joomla/joomla-cms/pull/3839/files#diff-3b9c9430ff5a041618b0075c9aefdec4)
- [libraries/cms/helper/media.php](https://github.com/joomla/joomla-cms/pull/3839/files#diff-01a503a61baecf6428f93553e180a067)

The rest considers com_media or may be added as pull requests to Joomla core (JImage::flip, added CSS, drag'n'drop upload).

I'm interested in decoupling for testing purposes. IMHO this component is too tightly integrated with core to decouple it in sense of removing from distribution.

Bakual commented:
Imho, you can't decouple com_media. It's an extension like com_categories which always should be present. Various 3rd party extension will also depend on it being present (and be it only by using the formfield).
Personally speaking, I'd say it's wasted time trying to decouple the media manager.

Fedik commented:
from the theory point of view, it could be similar to Captcha:
Joomla itself provide just the "media" field, that display the input depend from which "Media manager" are default.

piotr-cz commented:
My motivation in decoupling is to be able to install new media manager snapshot in J3.5 installation (before it's released in J3.6, if it is).

As I've written before, I consider media manager as a core component, like com_config and for sure it shouldn't be distributed separately but developed atomically.

brianteeman commented:
You can just download and install the entire branch

On 28 April 2016 at 14:06, Piotr <notifications@github.com> wrote:

> My motivation in decoupling is to be able to install new media manager
> snapshot in J3.5 installation (before it's released in J3.6, if it is).
>
> As I've written before, I consider media manager as a core component, like
> com_config and for sure it shouldn't be distributed separately but
> developed atomically.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/3839#issuecomment-215417029>
>



-- 
Brian Teeman
Co-founder Joomla! and OpenSourceMatters Inc.
http://brian.teeman.net/
